src/edu/washington/escience/myria/expression/evaluate/GenericEvaluator.java

```diff
@@ -121,27 +131,93 @@ public String getJavaExpressionWithAppend() {
+     */
+    public ReadableColumn getResults() {
+      return results;
+    }
+    /**
+     * This is useful for, e.g., constructing a {@link TupleBatch} from several evaluators.
+     * This will return null if the result was not generated from a single-valued expression.
+     *
+     * @return a {@link Column<?>} containing results from {@link #evaluateColumn}
+     */
+    public Column<?> getResultsAsColumn() {
+      if (results instanceof Column<?>) {
+        return (Column<?>) results;
+      } else {
+        return null;
+      }
+    }
+    /**
+     * @return a {@link Column<Integer>} containing result counts from {@link #evaluateColumn}
+     */
+    public Column<?> getResultCounts() {
+      return resultCounts;
+    }
+  }
+
+  /**
    * Evaluate an expression over an entire TupleBatch and return the column of results. This method cannot take state
    * into consideration.
    *
    * @param tb the tuples to be input to this expression
-   * @return a column containing the result of evaluating this expression on the entire TupleBatch
+   * @return an {@link EvaluatorResult} containing the results and result counts of evaluating this expression on the entire TupleBatch
    * @throws InvocationTargetException exception thrown from janino
    */
-  public Column<?> evaluateColumn(final TupleBatch tb) throws InvocationTargetException {
+  public EvaluatorResult evaluateColumn(final TupleBatch tb) throws InvocationTargetException {
+    // Optimization for result counts of single-valued expressions.
+    final Column<?> constCounts = new ConstantValueColumn(1, Type.INT_TYPE, tb.numTuples());
+    final WritableColumn countsWriter;
+    if (getExpression().isMultivalued()) {
+      countsWriter = ColumnFactory.allocateColumn(Type.INT_TYPE);
+    } else {
+      // For single-valued expressions, the Java expression will never attempt to write to `countsWriter`.
+      countsWriter = null;
+    }
     ExpressionOperator op = getExpression().getRootExpressionOperator();
-    /* This expression just copies an input column. */
+    // Critical optimization: return a zero-copy reference to a column referenced by a pure `VariableExpression`.
     if (isCopyFromInput()) {
-      return tb.getDataColumns().get(((VariableExpression) op).getColumnIdx());
+      return new EvaluatorResult(
+          tb.getDataColumns().get(((VariableExpression) op).getColumnIdx()), constCounts);
     }
-
-    Type type = getOutputType();
-
-    ColumnBuilder<?> ret = ColumnFactory.allocateColumn(type);
-    for (int row = 0; row < tb.numTuples(); ++row) {
-      /** We already have an object, so we're not using the wrong version of put. Remove the warning. */
-      eval(tb, row, ret, null);
+    // For multivalued expressions, we may get more than `TupleBatch.BATCH_SIZE` results,
+    // so we need to allocate a `TupleBuffer` instead of a `ColumnBuilder` for results.
+    // (That is why we need `ReadableColumn` rather than `Column` as the result type.)
```

Comment:
By calling `TupleBuffer.finalResult()` we will get the list of finalized tuple batches which consist of `Column<?>`. Finalizing makes sense since `resultsBuffer` will not be changed after `eval()`s. Then we can avoid the awkwardness of having `Column` and `ReadableColumn` together.